File: src/main/java/com/maxciv/infer/plugin/tree/TreeNodeFactory.java
package com.maxciv.infer.plugin.tree;

import com.maxciv.infer.plugin.process.report.InferReport;
import com.maxciv.infer.plugin.process.report.InferViolation;

import javax.swing.tree.DefaultMutableTreeNode;

/**
 * @author maxim.oleynik
 * @since 01.12.2018
 */
public final class TreeNodeFactory {

    private static TreeNodeFactory instance = new TreeNodeFactory();

    private TreeNodeFactory() {
    }

    public static TreeNodeFactory getInstance() {
        return instance;
    }

    public DefaultMutableTreeNode createNode(Object userObject) {
        if (userObject instanceof InferViolation) {
            return new ViolationNode((InferViolation) userObject);
        } else if (userObject instanceof InferReport) {
            return new RootNode((InferReport) userObject);
        } else if (userObject instanceof String) {
            return new StringNode((String) userObject);
        }
        return null;
    }

    public DefaultMutableTreeNode createNode(InferViolation violation) {
        return new ViolationNode(violation);
    }

    public DefaultMutableTreeNode createDefaultRootNode() {
        return new RootNode();
    }
}

File: src/main/kotlin/com/maxciv/infer/plugin/tree/TreeNodeFactory.kt
package com.maxciv.infer.plugin.tree

import com.maxciv.infer.plugin.process.report.InferReport
import com.maxciv.infer.plugin.process.report.InferViolation

import javax.swing.tree.DefaultMutableTreeNode

/**
 * @author maxim.oleynik
 * @since 01.12.2018
 */
class TreeNodeFactory private constructor() {

    fun createNode(userObject: Any): DefaultMutableTreeNode? {
        if (userObject is InferViolation) {
            return ViolationNode(userObject)
        } else if (userObject is InferReport) {
            return RootNode(userObject)
        } else if (userObject is String) {
            return StringNode(userObject)
        }
        return null
    }

    fun createNode(violation: InferViolation): DefaultMutableTreeNode {
        return ViolationNode(violation)
    }

    fun createDefaultRootNode(): DefaultMutableTreeNode {
        return RootNode()
    }

    companion object {
        val instance = TreeNodeFactory()
    }
}
